### ISSUE TITLE
Inconsistent results on every run with GPU

### ISSUE THREAD

**Issue by Ending2015a** [tensorflow/tensorflow#52680]
**System information**
- OS Platform and Distribution: Colab
- TensorFlow version: 2.6.0 GPU

**Describe the current behavior**
Hello, I'm using TensorFlow 2.4.0 with GPU support on my local machine to train a small CNN model. I can get the same training results when I train it on CPU mode if I fixed the random seed, but always get inconsistent results on GPU mode.

*So I test it on Colab, which has TensorFlow 2.6.0 installed in default, but still got inconsistent results on GPU mode.*

**Describe the expected behavior**
The results should be the same.

**[Contributing](https://www.tensorflow.org/community/contribute)**

- Do you want to contribute a PR? (yes/no): no
- Briefly describe your candidate solution(if contributing): x

**Standalone code to reproduce the issue**
Here is the test code
[Colab](https://colab.research.google.com/drive/1Xr_p9rsKM0bk3DDt9CL3L-dN7Vvz179Q?usp=sharing)
Run it for many times with GPU enabled, you will get different results on each run.

**Other info / logs*
I found that only when I use `relu`, `leaky_relu` or `tf.maximum` as the activation function the results change on every run. But if I use `sigmoid`, `tanh`, or other non-relu-like activation functions the results will remain the same.

**Comment by sachinprasadhs** [comment]
There was some changes made from Tensorflow 2.4 version which states as below.
`Optimizer.__init__ now accepts a gradient_aggregator to allow for customization of how gradients are aggregated across devices, as well as gradients_transformers to allow for custom gradient transformations (such as gradient clipping).`
May be this could be the reason, you can refer to [this](https://github.com/tensorflow/tensorflow/blob/master/RELEASE.md#tfkeras) 2.4 release note. Thanks.

**Comment by sachinprasadhs** [comment]
I'm able to see the same result in multiple runs using Tensorflow 2.7, please find the Gist [here](https://colab.research.google.com/gist/sachinprasadhs/5751a9c525624aa9132c3975d9dfba05/tensorflow-gradient-bug.ipynb). Thanks!

**Comment by Ending2015a** [comment]
@sachinprasadhs,

No, the loss is still changing in TF 2.7.
The `final sampled` is the training data sampled from numpy so it must be fixed.

![Screenshot from 2021-11-12 15-52-01](https://user-images.githubusercontent.com/18180004/141430348-27da70bb-72b5-41c9-9a30-370455e69e2c.png)

![Screenshot from 2021-11-12 15-49-32](https://user-images.githubusercontent.com/18180004/141430136-bc9484e6-2831-4dca-9b27-03b8b601e32c.png)

**Comment by sachinprasadhs** [comment]
I see it now, thanks for pointing it out, as you have mentioned that it happens only with certain activation functions, we may have to investigate on that first.
Since, development of keras moved to separate repository https://github.com/keras-team/keras/issues

Please post this issue on keras-team/keras repo.
To know more see;
https://discuss.tensorflow.org/t/keras-project-moved-to-new-repository-in-https-github-com-keras-team-keras/1999
Thank you!

**Comment by Ending2015a** [comment]
@sachinprasadhs, OK, I will post it to keras, thank you.

**Comment by tranvansang** [comment]
from TF 2.9, the following config works perfect, even with use_multithreading enabled

```python
import tensorflow as tf

tf.keras.utils.set_random_seed(1)
tf.config.experimental.enable_op_determinism()
```

https://stackoverflow.com/a/72377349/1398479
